executor/admin.go

```diff
@@ -138,3 +145,319 @@ func (e *CheckIndexRangeExec) constructIndexScanPB() *tipb.Executor {
+}
+
+func (e *RecoverIndexExec) extractIdxVals(row chunk.Row, idxVals []types.Datum) []types.Datum {
+	if idxVals == nil {
+		idxVals = make([]types.Datum, 0, row.Len()-1)
+	} else {
+		idxVals = idxVals[:0]
+	}
+
+	for i := 0; i < row.Len()-1; i++ {
+		colVal := row.GetDatum(i, &e.columns[i].FieldType)
+		idxVals = append(idxVals, *colVal.Copy())
+	}
+	return idxVals
+}
+
+type recoverRows struct {
+	handle  int64
+	idxVals []types.Datum
+	skip    bool
+}
+
+func (e *RecoverIndexExec) fetchRecoverRows(ctx context.Context, srcResult distsql.SelectResult, result *backfillResult) ([]recoverRows, error) {
+	e.recoverRows = e.recoverRows[:0]
+	handleIdx := len(e.columns) - 1
+	result.scanRowCount = 0
+Loop:
+	for {
+		err := srcResult.NextChunk(ctx, e.srcChunk)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		if e.srcChunk.NumRows() == 0 {
+			break
+		}
+		iter := chunk.NewIterator4Chunk(e.srcChunk)
+		for row := iter.Begin(); row != iter.End(); row = iter.Next() {
+			if result.scanRowCount >= int64(e.batchSize) {
+				break Loop
+			}
+			handle := row.GetInt64(handleIdx)
+			idxVals := e.extractIdxVals(row, e.idxValsBufs[result.scanRowCount])
+			e.idxValsBufs[result.scanRowCount] = idxVals
+			e.recoverRows = append(e.recoverRows, recoverRows{handle: handle, idxVals: idxVals, skip: false})
+			result.scanRowCount++
+			result.nextHandle = handle + 1
+		}
+	}
+
+	return e.recoverRows, nil
+}
+
+func (e *RecoverIndexExec) batchMarkDup(txn kv.Transaction, rows []recoverRows) error {
+	if len(rows) == 0 {
+		return nil
+	}
+
+	sc := e.ctx.GetSessionVars().StmtCtx
+	batchKeys := make([]kv.Key, 0, len(rows))
+	distinctFlags := make([]bool, len(rows))
+	for i, row := range rows {
+		idxKey, distinct, err := e.index.GenIndexKey(sc, row.idxVals, row.handle, e.idxKeyBufs[i])
+		if err != nil {
+			return errors.Trace(err)
+		}
+		e.idxKeyBufs[i] = idxKey
+
+		batchKeys = append(batchKeys, idxKey)
+		distinctFlags[i] = distinct
+	}
+
+	values, err := txn.GetSnapshot().BatchGet(batchKeys)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// 1. unique-key is duplicate and the handle is equal, skip it.
+	// 2. unique-key is duplicate and the handle is not equal, data is not consistent, log it and skip it.
+	// 3. non-unique-key is duplicate, skip it.
+	for i, key := range batchKeys {
+		if val, found := values[string(key)]; found {
+			if distinctFlags[i] {
+				handle, err1 := tables.DecodeHandle(val)
+				if err1 != nil {
+					return errors.Trace(err1)
+				}
+
+				if handle != rows[i].handle {
+					log.Warnf("[recover-index] The constraint of unique index:%v is broken, handle:%v is not equal handle:%v with idxKey:%v.",
+						e.index.Meta().Name.O, handle, rows[i].handle, key)
+				}
+			}
+			rows[i].skip = true
+		}
+	}
+	return nil
+}
+
+func (e *RecoverIndexExec) backfillIndexInTxn(ctx context.Context, txn kv.Transaction, startHandle int64) (result backfillResult, err error) {
+	result.nextHandle = startHandle
```

Comment:
Could we remove this line?

Reply:
no, if remove this, the result.nextHandle will be 0, is not correct.

Reply:
I think it's correct, because of `result.nextHandle` set in line351. But some cases the log may be not correct.
So there is no need to make any changes.
Ignore it.